### ISSUE TITLE
Desired behaviors for unbounded ranges

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#20896]
Traditionally, Chapel has defined unbounded ranges as being logically bounded by infinities, though this definition was very specific to integral ranges like `1..`.  In https://github.com/chapel-lang/chapel/issues/17084 it was pointed out that for unbounded ranges over a type that is inherently finite like enums or bools, it would make more sense to have them logically stop at the end of the legal values, and this was partially implemented in #19991 with some follow-ups being addressed in #20448.  For example, given an `enum color {red, orange, yellow, green, blue, indigo, violet};`, `for i in (color.red..)` will now iterate through all the colors and then terminate the loop (where before, it would run off the end of the loop).

This issue attempts to define what other behaviors we'd like to see from these unbounded ranges on finite types.  I'm particularly interested in the questions about first/last, high/low, highBound/lowBound behaviors.

  For example:
* iteration: The current behavior, of iterating up to the extreme value of the idxType, seems very reasonable and attractive to me
* slicing: This is a case where I think it makes sense to have unbounded ranges continue to behave as they currently do, which is to say, to inherit the array's bounds.  As an example, if I had an array defined over `red..green` in the above example and I sliced it with `orange..` I think I should get `orange..green` rather than getting an OOB error because we treat the unbounded range like `orange..violet`.
* first/last: I tend to think of these in terms of iteration, as in "What's the first value I'd iterate over?  What's the last value I'd iterate to?", so I'd be inclined to have these be defined for unbounded ranges to match the iteration decision above, where `orange..` would have a first of `orange` and a last of `violet`.
* [has][low|high]Bound: This is where things start to get tricky for me.  On one hand, I could imagine saying that "orange" and "violet" are the bounds of `orange..` but on the other hand, we call this a "boundedLow" range (or some equivalent if we end up renaming the enums).  So it might be a little weird to say "This is a low-bounded, high-unbounded range, but yes it has a high bound, and that high bound is violet" and I'd be inclined to say it does not have bounds without more rationale.  That said, I could be convinced I'm wrong here.
* low/high: And then this one is squarely in the middle of the previous two.  I think my inclination is to treat it more like first/last since it doesn't use the word "bound" and I tend to think of low/high as being like the non-directional versions of first/last...  But again, interested in other opinions.
* size: Similarly in a middle grey area for me.  Does it have a size because it has a first and last and can be iterated over?  Or does it not have one because its size is vague by virtue of being unbounded?  My inclination here would be to say that it does have a size, primarily because it feels nicer to give a numeric answer to size queries rather than to say "You can't do that!"
* domain/array declarations: I'm inclined to not support these for the time being—even though we arguably could—for the following reason:  I think that, ultimately, we may want to have a notion of unbounded domains (https://github.com/chapel-lang/chapel/issues/10595) where slicing with an unbounded domain would be similar to slicing with unbounded ranges; yet declaring an array over an unbounded domain would generally not be supported for unbounded integral domains.  Of course it could be for unbounded enum/bool domains, and we could implement that today, but I think that our approach would likely involve turning them into bounded domains (since we don't yet have unbounded domains), which would then result in the wrong slicing behavior.  That makes me inclined to put this off until #10595 is more generally supported.

**Comment by bmcdonald3** [comment]
I don't know that I have too much valuable input here, but I just wanted to note that, to me, it seems odd that `red..violet` is a different range than using any other value at either the low or high bound there, like, say, `orange..indigo`. Yes, these values are in the middle for how the enum was declared, but it just seems like what is being asked for is a range over the colors and I don't know that I would expect that to exclude any values without reading your comments above because I feel like the ordering of enums (or at least the ones that I've written) are generally arbitrary. In other words, I don't really see the connection for position of declaration to slices with enums with respect to ordering/bounding, but I can't say more there than that it feels odd, so may not be worth considering (and, also, this is just derailing the topic, so I wouldn't mind this being dismissed for the sake of progress).

**Comment by bradcray** [comment]
> but it just seems like what is being asked for is a range over the colors and I don't know that I would expect that to exclude any values without reading your comments above because I feel like the ordering of enums (or at least the ones that I've written) are generally arbitrary.

Yeah, that ship has long since sailed.  Ranges over enums are defined relative to the enums' symbols' ordering. as are comparison operators like `<`, `<=`, etc.  Presumably users would only be motivated to create a range over an enum when its symbols had a logical ordering and were declared to reflect it (E.g., S, M, L, XL).  See https://github.com/chapel-lang/chapel/issues/9117 for some of the public history here.

**Comment by vasslitvinov** [comment]
Fundamentally, I suggest interpreting various operations on ranges as operating on the represented sequence. For example, the equality operator `==` should compare the two operands' represented sequences. Which I believe it does in other contexts. For example, this should be true: `false.. == ..true`.

The following operations can be interpreted in this manner: iteration, first/last, size, ==.

For the operations that have "bound" in their name, I am in favor of leaving them boundedness-aware. For example, `(false..).hasHighBound()` is false. I am also game to avoid OOB errors while slicing with unbounded ranges, like we do with integer ranges. We need to have at least some functionality available to distinguish between `false..` and `false..true`.

low/high are a challenge here, as Brad discusses above. I was going to insist on treating them as properties on represented sequences. However, this does not work when the represented sequence is empty while the range is (at least semi-) bounded. I am on the fence here.

One approach we could take to refine the above is to think about the scenarios where the choice of behavior will make a difference in user code.

**Comment by vasslitvinov** [comment]
Based on today's group discussion, I have an update to the above:
* first and last should be defined because hasFirst() and hasLast() gotta return true because the represented sequences have those
* let's make `low`/`high` cause compile-time errors because (a) my intuition tells me they should be defined, however (b) others disagree with that, so (c) it is easier to go from error to non-error in the future when there is a worthy use case

**Comment by benharsh** [comment]
Based on our offline discussion, I tend to lean more towards the "restrictive" approach. I think it's probably safer and more useful to force/encourage users to consider the boundedness of the ranges they are working with, particularly in the case of generic arguments with a range type (e.g. ``range(?)``).

I also think it might be convenient to have a method on ranges that can return a bounded value based on the min/max of the index type and whether the original value was bounded. It would also be nice to have type functions on a range to get the min/max of the possible values for an index type.

**Comment by bradcray** [comment]
I think my next step here will be to explore my current PR (that found that it could support all the iterators cleanly and easily by supporting these queries on unbounded enum/bool ranges) to see what the code would look like if all/some of the queries were unsupported.  To an extent, this should also give a sense of what the impact on user code will be / what utilities the user code would need to have, etc.

**Comment by vasslitvinov** [comment]
@bradcray and I worked out the following conclusions. They are based off the existing iteration behavior, which is always "as if the range is bounded" when the range's idxType is finite.

| feature | current behavior | behavior for 2.0 |
| --- | --- | --- |
| first, last | as if bounded | as if bounded; unstable |
| == | as if unbounded | as if bounded; unstable |
| hasHighBound, high[Bound] | as if unbounded | same |
| hasLowBound, low[Bound] | " | " |
| size | as if unbounded | same |
| domains, arrays | as if unbounded | same |

#### Rationale

* For `first` and `last`, the name suggests "first/last in the sequence." Given that the sequence is finite if idxType is finite, it makes sense to `first` and `last` defined as if the range is bounded, similarly to iteration. To be on the safe side, we will mark this behavior as unstable.

* For `==`, we have thought of it as comparing the sequences represented by the two ranges. For example, `(1..8 by 2 align 0) == (2..9 by 2)` is true, despite the two ranges having different bounds. For finite idxTypes the represented sequence of an unbounded range matches that for a similar bounded range. For example, `false..` and `..: range(bool)` have the bounded sequence of `false..true`. Therefore we want to treat unbounded ranges as bounded for the purpose of `==`.

For those features that treat the range as unbounded, it is a conservative choice:
* For `hasHighBound`, `highBound`, and `high`, and similarly `...low...`, it restricts what users are allowed to do. For example, it disallowes queries of the high bound in this case. Note that switching this to a less conserviative approach would not be backwards compatible because it would entail changing the return value of hasHighBound().

* For `size` and [declaring] domains+arrays, this choice allows us to enable them later, which would be backward-compatible.
